Service workers

rlr opened the pull request:
here it is again. r?

rlr committed:
Service worker.

rlr commented:
I am not sure what's up with the tests. Are they supposed to work in this branch?

```
$ npm test

> newtab@0.0.0 test /Users/rlr/dev/remote-newtab
> npm run karma


> newtab@0.0.0 karma /Users/rlr/dev/remote-newtab
> karma start --single-run

sh: karma: command not found

npm ERR! Darwin 15.2.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "karma"
npm ERR! node v4.2.1
npm ERR! npm  v2.14.7
npm ERR! file sh
npm ERR! code ELIFECYCLE
npm ERR! errno ENOENT
npm ERR! syscall spawn
npm ERR! newtab@0.0.0 karma: `karma start --single-run`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the newtab@0.0.0 karma script 'karma start --single-run'.
npm ERR! This is most likely a problem with the newtab package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     karma start --single-run
npm ERR! You can get their info via:
npm ERR!     npm owner ls newtab
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/rlr/dev/remote-newtab/npm-debug.log
npm ERR! Test failed.  See above for more details.
```

rlr commented on sw.js:
```diff
@@ -0,0 +1,82 @@
+        './img/icon-gear.svg',
+        './img/magnifier.png',
+        './img/pin.svg',
+        './img/x.svg'
```
I need to change this to use the list in `src/js/mainSiteURLs.json`

rlr commented:
^ I updated the service worker to use the list of static files from mainSiteURLs.json

k88hudson commented on package.json:
```diff
@@ -4,9 +4,9 @@
   "main": "src/main.js",
   "scripts": {
-    "setup": "npm run clean && npm run build:static",
+    "setup": "npm run clean && npm run build:static && npm run build:offline",
```
This `build:offline` needs to run after `build:js` and `build:css` because the files will not have been generated yet, no?

rlr replied:
hmmm, this is true. I thought it was looking at `src/static`. Maybe I should change it to that? I need the file to be there during dev.

rlr replied:
actually. it needs to look at www. So maybe we have to do a full build during setup :/

k88hudson commented on src/lib/registerServiceWorker.js:
```diff
@@ -0,0 +1,16 @@
+  if ('serviceWorker' in navigator) {
+    navigator.serviceWorker.register('./sw.js', {scope: './'}).then(function (reg) {
+      if (reg.installing) {
+        console.log('Service worker installing');
```
There is a `lib/log` function if you want to keep these but avoid lint warnings

rlr committed:
move build:offline back

rlr committed:
fix logging

rlr commented:
ok. i moved the build:offline back. I think it's fine.. Only when debugging the service worker would we need to do a full build before starting the dev server. Fixed the logging too

k88hudson commented on src/static/sw.js:
```diff
@@ -0,0 +1,83 @@
+  // synchronously during the event dispatch.  In the future we should
+  // be able to move the evt.waitUntil() down to directly where we
+  // need it in the respondWith() async handling.
+  var waitUntilResolve;
```
use let

k88hudson commented on src/static/sw.js:
```diff
@@ -0,0 +1,83 @@
+    waitUntilResolve = resolve;
+  }));
+
+  var cache;
```
same thing here

rlr committed:
s/var/let/

k88hudson commented:
YAY! this is looking awesome! R+

k88hudson commented on src/static/sw.js:
```diff
@@ -0,0 +1,83 @@
+  }));
+
+  var cache;
+  evt.respondWith(caches.open(CACHE_NAME).then(function(c) {
```
If I were marcos I'd be all like, "use a generator!" but since I'm not marcos I won't :stuck_out_tongue_winking_eye:
